app/src/main/java/org/connectbot/TerminalView.java

```diff
@@ -150,6 +171,282 @@ public TerminalView(Context context, TerminalBridge bridge) {
+				public boolean onSingleTapConfirmed(MotionEvent e) {
+					viewPager.performClick();
+					return super.onSingleTapConfirmed(e);
+				}
+			});
+		}
+	}
+
+	@TargetApi(11)
+	private void initSelectionCallback() {
+		this.setCustomSelectionActionModeCallback(new ActionMode.Callback() {
+			private static final int PASTE = 0;
+
+			@Override
+			public boolean onCreateActionMode(ActionMode mode, Menu menu) {
+				TerminalView.this.selectionActionMode = mode;
+
+				menu.add(0, PASTE, 2, "Paste")
+						.setIcon(R.drawable.ic_action_paste)
+						.setShowAsAction(MenuItem.SHOW_AS_ACTION_WITH_TEXT | MenuItem.SHOW_AS_ACTION_IF_ROOM);
+
+				return true;
+			}
+
+			@Override
+			public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
+				if (item.getItemId() == PASTE) {
+					String clip = clipboard.getText().toString();
+					TerminalView.this.bridge.injectString(clip);
+					mode.finish();
+					return true;
+				}
+
+				return false;
+			}
+
+			@Override
+			public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
+				return false;
+			}
+
+			@Override
+			public void onDestroyActionMode(ActionMode mode) {
+			}
+		});
+	}
+
+	public void copyCurrentSelectionToClipboard() {
+		ClipboardManager clipboard =
+				(ClipboardManager) TerminalView.this.context.getSystemService(Context.CLIPBOARD_SERVICE);
+		if (currentSelection.length() != 0) {
+			clipboard.setText(currentSelection);
+		}
+
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB &&
+			selectionActionMode != null) {
+			selectionActionMode.finish();
+			selectionActionMode = null;
+		}
+	}
+
+	@Override
+	protected void onSelectionChanged(int selStart, int selEnd) {
+		if (selStart <= selEnd) {
+			currentSelection = getText().toString().substring(selStart, selEnd);
+		}
+		super.onSelectionChanged(selStart, selEnd);
+	}
+
+	@Override
+	public boolean performLongClick() {
+		refreshTextFromBuffer();
+		return super.performLongClick();
+	}
+
+	@Override
+	public boolean onTouchEvent(MotionEvent event) {
+		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB) {
+			return false;
+		}
+
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH &&
+				MotionEventCompat.getSource(event) == InputDevice.SOURCE_MOUSE) {
+			if (onMouseEvent(event, bridge)) {
+				return true;
+			}
+			viewPager.setPagingEnabled(true);
+		}
+
+		super.onTouchEvent(event);
+		if (gestureDetector != null) {
+			gestureDetector.onTouchEvent(event);
+		}
+
+		return true;
+	}
+
+	/**
+	 * @param event
+	 * @param bridge
+	 * @return True if the event is handled.
+	 */
+	@TargetApi(14)
+	private boolean onMouseEvent(MotionEvent event, TerminalBridge bridge) {
+		int row = (int) Math.floor(event.getY() / bridge.charHeight);
+		int col = (int) Math.floor(event.getX() / bridge.charWidth);
+		int meta = event.getMetaState();
+		boolean shiftOn = (meta & KeyEvent.META_SHIFT_ON) != 0;
+		boolean mouseReport = ((vt320) bridge.buffer).isMouseReportEnabled();
```

Comment:
Hmm I guess there are a few options. What Kyle mentions is probably the easiest since the bridge never changes in TerminalView. It's a bit risky because if anyone ever adds code to change the bridge or change the buffer inside the bridge, this instance variable will also need to be updated. Another alternatives would be to either add the mouse functions to the VDUBuffer (not sure they really belong there) or switch the type of TerminalBridge.buffer to a vt320, which exposes a much bigger interface, but we're working around the interface anyway here...